File: haskell/aoc2017/src/Day09.hs
module Day09
  ( part1
  , part2
  ) where

parseStream :: String -> Int
parseStream input = parseStream_ input [] False 0

parseStream_ :: String -> [Int] -> Bool -> Int -> Int
parseStream_ [] l _ _ = sum l
parseStream_ (x:xs) l g i =
  case x of
    '!' -> parseStream_ (tail xs) l g i
    '{' ->
      if g
        then parseStream_ xs l g i
        else parseStream_ xs l g (i + 1)
    '<' -> parseStream_ xs l True i
    '>' -> parseStream_ xs l False i
    '}' ->
      if g
        then parseStream_ xs l g i
        else parseStream_ xs (i : l) g (i - 1)
    _ -> parseStream_ xs l g i

removeGarbage :: String -> Int
removeGarbage input = removeGarbage_ input 0 False

removeGarbage_ :: String -> Int -> Bool -> Int
removeGarbage_ [] s _ = s
removeGarbage_ (x:xs) s g =
  case x of
    '!' -> removeGarbage_ (tail xs) s g
    '<' ->
      removeGarbage_
        xs
        (if g
           then s + 1
           else s)
        True
    '>' -> removeGarbage_ xs s False
    _ ->
      removeGarbage_
        xs
        (if g
           then s + 1
           else s)
        g

-- export
part1 :: String -> String
part1 input = show $ parseStream input

part2 :: String -> String
part2 input = show $ removeGarbage input

File: aoc_ruby/lib/2015/day_09.rb
# frozen_string_literal: true

module AoC2015
  class Day09
    def part1(input)
      find(input, :min)
    end

    def part2(input)
      find(input, :max)
    end

    private

    def find(input, f)
      distance_map = distance_map(input)
      routes = distance_map.keys.flatten.uniq.permutation
      distances = routes.map { |route| route_distance(route, distance_map) }
      distances.send(f)
    end

    def route_distance(route, distance_map)
      route.each_cons(2).sum { |path| distance_map[path.sort] }
    end

    def distance_map(input)
      input
        .lines
        .map { |line| line.split(/(.+) to (.+) = (\d+)/).drop(1) }
        .each_with_object({}) { |(a, b, d), m| m[[a, b].sort] = d.to_i }
    end
  end
end
